### PR TITLE
[Incremental] Skip post-compile jobs only if no compiles ran, and all the outputs are up-to-date.

### PR DESCRIPTION
**Pull request by davidungar**
If the merge-modules step fails or is cancelled, all the compile steps may be skipped, yet the post-compile jobs still must run.
Improve the logic that was skipping the post-compile jobs whenever there are no compile jobs.
When the compile jobs would otherwise be skipped, check their outputs against the time the previous build finished.
In order to perform the check, enhance the build record to store the end time as well as the start time of the prior build.

rdar://76127929

### CODE REVIEW

**Comment by davidungar** [comment]
@swift-ci please test

**Comment by davidungar** [comment]
@swift-ci please test

File: Sources/SwiftDriver/IncrementalCompilation/IncrementalCompilationState.swift [hunk 1]
```diff
@@ -46,6 +46,14 @@ public final class IncrementalCompilationState {
   /// Jobs to run *after* the last compile, for instance, link-editing.
   public let jobsAfterCompiles: [Job]
 
+  /// Can compare input mod times against this.
+  private let buildStartTime: Date
+
+  /// In order to decide while postCompile jobs to run, need to compile their outputs against the build time
```
Comment (cltnschlosser) [inline Sources/SwiftDriver/IncrementalCompilation/IncrementalCompilationState.swift:52]:
```suggestion
  /// In order to decide while postCompile jobs to run, need to compare their outputs against the build time
```

File: Sources/SwiftDriver/IncrementalCompilation/IncrementalCompilationState.swift [hunk 2]
```diff
@@ -290,6 +305,14 @@ extension IncrementalCompilationState {
     return try driver.formBatchedJobs(unbatched, showJobLifecycle: driver.showJobLifecycle)
   }
 }
+// MARK: - Scheduling post-compile jobs
+extension IncrementalCompilationState {
+  public func canSkipPostCompile(job: Job) -> Bool {
+    job.outputs.allSatisfy {output in
+      let fileModTime = (try? fileSystem.lastModificationTime(for: output.file)) ?? .distantFuture
+      return fileModTime < buildEndTime}
```
Comment (cltnschlosser) [inline Sources/SwiftDriver/IncrementalCompilation/IncrementalCompilationState.swift:313]: `<=`?

File: Sources/SwiftDriver/IncrementalCompilation/IncrementalCompilationState.swift [hunk 3]
```diff
@@ -290,6 +305,14 @@ extension IncrementalCompilationState {
     return try driver.formBatchedJobs(unbatched, showJobLifecycle: driver.showJobLifecycle)
   }
 }
+// MARK: - Scheduling post-compile jobs
+extension IncrementalCompilationState {
+  public func canSkipPostCompile(job: Job) -> Bool {
+    job.outputs.allSatisfy {output in
+      let fileModTime = (try? fileSystem.lastModificationTime(for: output.file)) ?? .distantFuture
```
Comment (cltnschlosser) [inline Sources/SwiftDriver/IncrementalCompilation/IncrementalCompilationState.swift:312]: Is this the function that just does stat?

**Review by cltnschlosser — APPROVED** [review]

**Comment by davidungar** [comment]
FYI, I plan to add a regression test before landing this PR.

**Comment by davidungar** [comment]
@cltnschlosser Thank you for such a quick review! I'll look at your concerns today.

File: Sources/SwiftDriver/IncrementalCompilation/IncrementalCompilationState.swift (hunk 1 above)
Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/IncrementalCompilationState.swift:52]: Good catch! A thinko on my part.

File: Sources/SwiftDriver/IncrementalCompilation/IncrementalCompilationState.swift (hunk 2 above)
Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/IncrementalCompilationState.swift:313]: Hmmm.... Good question. `<` counts on there being a little extra time for the driver to finish up... Sure, `<=` might be better. Thanks for asking the question.

File: Sources/SwiftDriver/IncrementalCompilation/IncrementalCompilationState.swift (hunk 3 above)
Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/IncrementalCompilationState.swift:312]: Not sure what you are asking. If you mean to ask why `canSkipPostCompile` is a separate function, it is because I have found it helpful to use "intention-revealing names", factoring functions where needed: http://wiki.c2.com/?IntentionRevealingNames .

OTOH, if you are asking about `fileSystem.lastModificationTime`, the answer is "yes".

Let me know if I misunderstood. TIA!

**Comment by davidungar** [comment]
@swift-ci please test

File: Sources/SwiftDriver/IncrementalCompilation/IncrementalCompilationState.swift (hunk 3 above)
Comment (cltnschlosser) [reply Sources/SwiftDriver/IncrementalCompilation/IncrementalCompilationState.swift:312]: Yeah the lastModificationTime, I was just double checking that this was the new function that calls stat directly. It’s much faster than the other one that did something like getFileInfo().modTime

Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/IncrementalCompilationState.swift:312]: Thanks for clarifying.

**Comment by davidungar** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
Err, can we change the format of the build record? If the user needs to fall back/upgrade to or from the old and new drivers, will this break anything?

**Comment by davidungar** [comment]
> Err, can we change the format of the build record? If the user needs to fall back/upgrade to or from the old and new drivers, will this break anything?

Thanks for taking a look at this PR. I'm not sure I understand. The new build end time has a new section name, and the code should accept either the old name for the old time or the new name for the old time. At least, that's the code I thought I wrote.
If the user upgrades and there are files to be compiled, it should just work as before, writing out the more complete build record.
If the user upgrades and there are no files to be compiled, the post-compile jobs might run one time, even though they are not needed.
If the user downgrades, the read of the build record start date will fail, and everything will be recompiled on the first post-downgrade compile.
The tradeoff here is between using a clearer name for the start date in the build record, and avoiding the recompilation-when-downgrading penalty. I had chosen the former, but did you mean to argue for the latter? If so, what is the rationale? I'm open to it.

**Comment by CodaFi** [comment]
The question is more base than that: This new format appears to replace a key from the legacy driver’s build record format outright. So, if I use the new driver and need to downgrade then the old driver will be missing information. Is the legacy driver tolerant of that use case? Similarly if I upgrade, the new driver will not have these new keys to read.

**Comment by davidungar** [comment]
> The question is more base than that: This new format appears to replace a key from the legacy driver’s build record format outright. So, if I use the new driver and need to downgrade then the old driver will be missing information. Is the legacy driver tolerant of that use case? Similarly if I upgrade, the new driver will not have these new keys to read.
> 
> If we are unsure of the answer to this question, I would prefer we make only additive changes to the build record format.

Thanks for explaining. This sort of issue is exactly why I like to reviews! I just performed a few experiments: Tried an upgrade from a main swift-driver to this one, a downgrade from this one to the main, and also a downgrade from this one to the legacy driver. I tried these transitions in one of two scenarios, with "build_time" changed to "build_start_time" and with "build_time" left alone.

Surprisingly, it turned out that there was no benefit to the latter scenario. Both the legacy driver and the main (status quo) swift driver reject the build record when there is a new field (in this case "build_end_time") added to it. I conclude that the better option is to go with the former scenario. We might as well rename the field since the added field will cause rejection anyway.

To summarize the results:

Upgrading from the main swift driver to this PR:
- Works fine

Downgrading from this PR to a main swift driver: 
- "Incremental compilation has been disabled due to malformed build record file Unexpected key 'build_start_time'"
- Subseqenent compilations work fine

Downgrading from this PR or from main swift driver to legacy driver:
- "Incremental compilation has been disabled, because different arguments were passed to the compiler."

Thanks for asking the question. I don't see any better solution than this PR as it stands. Do you?

**Review by CodaFi — APPROVED** [review]

**Comment by davidungar** [comment]
@swift-ci please test

**Comment by davidungar** [comment]
Also, cleaned up the commit history
